lib/benchmark/benchmark.go

```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		exitSignals := make(chan os.Signal, 1)
+		signal.Notify(exitSignals, syscall.SIGTERM, syscall.SIGINT)
+		defer signal.Stop(exitSignals)
+		sig := <-exitSignals
+		logrus.Debugf("signal: %v", sig)
+		cancel()
+	}()
+	var results []Result
+	sleep := false
+	for {
+		if sleep {
+			select {
+			case <-time.After(pauseTimeBetweenBenchmarks):
+			case <-ctx.Done():
+				return results, trace.ConnectionProblem(ctx.Err(), "context canceled or timed out")
+			}
+		}
+		benchmarkC := lg.GetBenchmark()
+		if benchmarkC == nil {
+			break
+		}
+		result, err := benchmarkC.Benchmark(ctx, tc)
+		if err != nil {
+			return results, trace.Wrap(err)
+		}
+		results = append(results, result)
+		fmt.Printf("current generation requests: %v, duration: %v\n", result.RequestsOriginated, result.Duration)
+		sleep = true
+	}
+	return results, nil
+}
+
+// ExportLatencyProfile exports the latency profile and returns the path as a string if no errors
+func ExportLatencyProfile(path string, h *hdrhistogram.Histogram, ticks int32, valueScale float64) (string, error) {
+	timeStamp := time.Now().Format("2006-01-02_15:04:05")
+	suffix := fmt.Sprintf("latency_profile_%s.txt", timeStamp)
+	if path != "." {
+		if err := os.MkdirAll(path, 0700); err != nil {
+			return "", trace.Wrap(err)
+		}
+	}
+	fullPath := filepath.Join(path, suffix)
+	fo, err := os.Create(fullPath)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if _, err := h.PercentilesPrint(fo, ticks, valueScale); err != nil {
+		if err := fo.Close(); err != nil {
+
+			logrus.WithError(err).Warningf("failed to close file")
+		}
+		return "", trace.Wrap(err)
+	}
+
+	if err := fo.Close(); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return fo.Name(), nil
+}
+
+// Benchmark connects to remote server and executes requests in parallel according
+// to benchmark spec. It returns benchmark result when completed.
+// This is a blocking function that can be cancelled via context argument.
+func (c *Config) Benchmark(ctx context.Context, tc *client.TeleportClient) (Result, error) {
+	tc.Stdout = ioutil.Discard
+	tc.Stderr = ioutil.Discard
+	tc.Stdin = &bytes.Buffer{}
+	var delay time.Duration
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	requestsC := make(chan benchMeasure)
+	resultC := make(chan benchMeasure)
+
+	if c.Rate <= 0 {
+		return Result{}, errors.New("rate must be greater than 0")
+	}
+
+	go func() {
+		interval := time.Duration(1 / float64(c.Rate) * float64(time.Second))
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+		start := time.Now()
+		for {
+			select {
+			case <-ticker.C:
+				// ticker makes its first tick after the given duration, not immediately
+				// this sets the send measure ResponseStart time accurately
+				delay = delay + interval
+				t := start.Add(delay)
+				measure := benchMeasure{
```

Comment:
No, what I meant was something along the lines of:
```go
newMeasure := func(ts time.Time) benchMeasure {
	return benchMeasure{
		ResponseStart: ts,
		command:       c.Command,
		client:        tc,
		interactive:   c.Interactive,
	}
}

go func() {
	interval := time.Duration(1 / float64(c.Rate) * float64(time.Second))
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	go work(ctx, newMeasure(time.Now()), resultC)
	for {
		select {
		case ts := <-ticker.C:
			go work(ctx, newMeasure(ts), resultC)
		case <-ctx.Done():
			close(requestsC)
			return
		}
	}
}()
```

Reply:
I see what you mean, this would work because this sends one request right when the ticker starts and then just sets the ticket time as the start in the measure. yes , this is much simpler.